### PR TITLE
bug(blockchain): make block finalization wait on sidecar availability

### PR DESCRIPTION
**Pull request by abi87**
When we finalize a block, we verify that its associated sidecar has been stored.
However block and sidecar finalization are handled concurrently, so we need wait for the blob to be persisted before checking.
This PR solves the issue

## Release Notes

- **New Features**
	- Introduced asynchronous verification for blob availability during block processing, improving responsiveness.
	- Added new channels for managing blob sidecar finalization events.
	- Implemented a new event handling capability for `BlobSidecarsFinalized`.
	- Added a constant for `BlobSidecarsFinalized` to improve event identification.
	- Enhanced `ProcessSidecars` method to return additional contextual information (slot number) alongside error handling.

- **Bug Fixes**
	- Enhanced error handling for event publishing, ensuring robustness in the event handling process.
  
- **Documentation**
	- Updated comments for clarity in the codebase.

### CODE REVIEW

File: mod/beacon/blockchain/service.go [hunk 1]
```diff
@@ -80,6 +80,12 @@ type Service[
 	// forceStartupSyncOnce is used to force a sync of the startup head.
 	forceStartupSyncOnce *sync.Once
 
+	// blobFinalized is used to verify blob sidecar availability upon
+	// block finalization
+	blobFinalized chan struct{}
```
Comment (abi87) [inline mod/beacon/blockchain/service.go:85]: @itsdevbear we may want to harden this and pass some kind of blob data to enhance verification. Not sure what. You got suggestions?

File: mod/beacon/blockchain/process.go [hunk 2]
```diff
@@ -63,13 +64,17 @@
...
-		ctx, blk.GetSlot(), blk.GetBody(),
-	) {
-		return nil, ErrDataNotAvailable
+	// To complete block finalization, we wait for the associated blob
+	// to be persisted and verify it's available. Once that is done we
+	// complete block finalization.
+	g, _ := errgroup.WithContext(ctx)
+	g.Go(func() error {
+		return s.verifyFinalBlobAvailability(ctx, blk)
+	})
```
[the remark this answers is not in the document]
Comment (calbera) [reply mod/beacon/blockchain/process.go:73]: +1 should use the context from `errgroup` here

[the remark this answers is not in the document]
Comment (calbera) [reply mod/beacon/blockchain/process.go:73]: Wait why are we using errgroup here for 1 task? Usually err group helps for multiple concurrent tasks. Also we just end up waiting on its result anyways below, so it seems we can just call `verifyFinalBlobAvailability` in the same thread?

File: mod/da/pkg/da/service.go [hunk 3]
```diff
@@ -132,8 +132,16 @@ func (s *Service[_, BlobSidecarsT]) handleFinalSidecarsReceived(
...
 		s.logger.Error(
 			"Failed to process blob sidecars",
-			"error",
-			err,
+			"error", err,
+		)
+		return
+	}
+
+	event := async.NewEvent(msg.Context(), async.BlobSidecarsFinalized, struct{}{})
```
Comment (calbera) [inline mod/da/pkg/da/service.go:140]: might be worth adding the blk slot num in the event data here

File: mod/beacon/blockchain/service.go [hunk 4]
```diff
@@ -313,3 +329,27 @@ func (s *Service[
...
+	// If the blobs needed to process the block are not available, we
+	// return an error. It is safe to use the slot off of the beacon block
+	// since it has been verified as correct already.
+	if !s.storageBackend.AvailabilityStore().IsDataAvailable(
+		ctx, blk.GetSlot(), blk.GetBody(),
+	) {
+		return ErrDataNotAvailable
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (calbera) [reply mod/beacon/blockchain/service.go:355]: would agree, might need to check that the blk.GetSlot() <=  number of blobFinalized event

[the remark this answers is not in the document]
Comment (calbera) [reply mod/beacon/blockchain/service.go:359]: Would need to prune this map then

File: mod/beacon/blockchain/service.go (hunk 1 above)
Comment (calbera) [reply mod/beacon/blockchain/service.go:86]: I think even making the data here be the slot number for which the blob belongs to is sufficient.

**Comment by abi87** [comment]
Solved by https://github.com/berachain/beacon-kit/commit/711ce6fd5efcde882dda9b40668adbce1c0af875.
We removed pub/sub system and ensured that when we finalize, sidecars are finalized before blocks

**Pull request closed without merging**